src/wp-includes/class-wp-token-map.php

```diff
@@ -0,0 +1,740 @@
+ * continuation bytes.
+ *
+ * Since it has not been shown during the development of this class that longer strings are required, this
+ * update is deferred until such a need is clear.
+ *
+ * @since 6.6.0
+ */
+class WP_Token_Map {
+	/**
+	 * Maximum length for each key and each transformed value in the table (in bytes).
+	 *
+	 * @since 6.6.0
+	 */
+	const MAX_LENGTH = 256;
+
+	/**
+	 * How many bytes of each key are used to form a group key for lookup.
+	 * This also determines whether a word is considered short or long.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var int
+	 */
+	private $key_length = 2;
+
+	/**
+	 * Stores an optimized form of the word set, where words are grouped
+	 * by a prefix of the `$key_length` and then collapsed into a string.
+	 *
+	 * In each group, the keys and lookups form a packed data structure.
+	 * The keys in the string are stripped of their "group key," which is
+	 * the prefix of length `$this->key_length` shared by all of the items
+	 * in the group. Each word in the string is prefixed by a single byte
+	 * whose raw unsigned integer value represents how many bytes follow.
+	 *
+	 *     ┌────────────────┬───────────────┬─────────────────┬────────┐
+	 *     │ Length of rest │ Rest of key   │ Length of value │ Value  │
+	 *     │ of key (bytes) │               │ (bytes)         │        │
+	 *     ├────────────────┼───────────────┼─────────────────┼────────┤
+	 *     │ 0x08           │ nterDot;      │ 0x02            │ ·      │
+	 *     └────────────────┴───────────────┴─────────────────┴────────┘
+	 *
+	 * In this example, the key `CenterDot;` has a group key `Ce`, leaving
+	 * eight bytes for the rest of the key, `nterDot;`, and two bytes for
+	 * the transformed value `·` (or U+B7 or "\xC2\xB7").
+	 *
+	 * Example:
+	 *
+	 *    // Stores array( 'CenterDot;' => '·', 'Cedilla;' => '¸' ).
+	 *    $groups      = "Ce\x00";
+	 *    $large_words = array( "\x08nterDot;\x02·\x06dilla;\x02¸" )
+	 *
+	 * The prefixes appear in the `$groups` string, each followed by a null
+	 * byte. This makes for quick lookup of where in the group string the key
+	 * is found, and then a simple division converts that offset into the index
+	 * in the `$large_words` array where the group string is to be found.
+	 *
+	 * This lookup data structure is designed to optimize cache locality and
+	 * minimize indirect memory reads when matching strings in the set.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var array
+	 */
+	private $large_words = array();
+
+	/**
+	 * Stores the group keys for sequential string lookup.
+	 *
+	 * The offset into this string where the group key appears corresponds with the index
+	 * into the group array where the rest of the group string appears. This is an optimization
+	 * to improve cache locality while searching and minimize indirect memory accesses.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var string
+	 */
+	private $groups = '';
+
+	/**
+	 * Stores an optimized row of small words, where every entry is
```

Comment:
I'd like to see the explicit rule for large vs. small words in their docs:

> …where small words are strings whose strlen is less than or equal to key_length

(that's almost certainly wrong, just an example)

Reply:
added a class-level description in [31c57b0](https://github.com/WordPress/wordpress-develop/pull/5373/commits/31c57b0871a193fdf18a6ae870cb32b0750799a2)